### PR TITLE
GH-101673: Fix pdb bug that local variable changes lost after longlist

### PR DESCRIPTION
**Pull request by gaogaotiantian**
The bug is described in #101673. 

Using `f_code.co_name` to check whether it's a frame module avoids reading `f_local` whose side effect reverts local variable changes in pdb.

* Issue: gh-101673

### CODE REVIEW

**Comment by ortem** [comment]
@iritkatriel I am wondering if you could be an appropriate reviewer for this. Could you please take a look?

**Comment by iritkatriel** [comment]
Thanks for adding the test. Before merging this I would need to spend some time to understand why this happened and to convince myself that this is the right solution.  If someone else reviews it before I get around to it that's obviously fine.

**Comment by ortem** [comment]
@iritkatriel Thank you for your quick response.

I have investigated the issue further and discovered that the problematic `obj.f_globals is obj.f_locals` check was introduced in https://github.com/python/cpython/commit/5ed2b5a92ab4f5c4c8ad77ec95a4f7add3083432 to work around the limitation of `inspect.getsourcelines`, as the latter did not check that the given frame is a module-level frame.

However, `inspect.getsourcelines` was independently corrected in https://github.com/python/cpython/commit/91cb298f811961277fd4cc4a32211899d48bedcb using the same `object.f_code.co_name == "<module>"` technique.

So I believe that we should consider replacing the custom `pdb.getsourcelines` function with direct `inspect.getsourcelines` invocations. After making this change, the updated `test_pdb` still passes. @gaogaotiantian WDYT?

**Comment by gaogaotiantian** [comment]
Yeah I think that’s the way to go. The original customized pdb version was to avoid module level frame issue and it would be nice to not to keep multiple identical code pieces around. The functionality seems to be the same. I’ll make the changes to use inspect version directly and remove the pdb version.

**Comment by gaogaotiantian** [comment]
As @ortem said, the reason we had a separate `getsourcelines` in pdb was because the `inspect` version did not deal with the module level frame. It does now so we can just switch back to `inspect` version and kept one less duplicated code piece. I did include the regression test for the issue.

File: Misc/NEWS.d/next/Library/2023-02-09-19-40-41.gh-issue-101673.mX-Ppq.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+Fix a pdb bug where ``ll`` clears the changes to local variables.
```
Comment (iritkatriel) [inline Misc/NEWS.d/next/Library/2023-02-09-19-40-41.gh-issue-101673.mX-Ppq.rst]:
```suggestion
Fix a :mod:`pdb` bug where ``ll`` clears the changes to local variables.
```

**Review by iritkatriel — APPROVED** [review]
Makes sense. Thank you!

**Pull request merged**